File: core/landmarks.py
"""Frame model, rolling buffer, and shoulder-width normalization.

This is the data backbone every later phase builds on:
  - Hand / Frame: one timestamped sample (21 landmarks/hand + pose shoulder points), stored in
    PIXEL coordinates so x and y share a unit regardless of webcam aspect ratio.
  - RollingBuffer: a fixed ~1.5-2s time window of Frames — the ONLY thing movement checks read.
  - normalized_distance(): every distance expressed as a ratio of shoulder width, so thresholds
    don't break when the user sits closer to or further from the camera.

Frames are JSON-serializable (to_dict/from_dict) so the Phase 4 recorder can save fixtures and
the confusor tests can replay them.
"""
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from typing import Iterator, Optional

import numpy as np

# --- MediaPipe hand landmark indices (21-point model) ---
WRIST = 0
THUMB_TIP = 4
INDEX_MCP, INDEX_TIP = 5, 8
MIDDLE_MCP, MIDDLE_TIP = 9, 12
RING_MCP, RING_TIP = 13, 16
PINKY_MCP, PINKY_TIP = 17, 20
FINGERTIPS = (THUMB_TIP, INDEX_TIP, MIDDLE_TIP, RING_TIP, PINKY_TIP)
MCPS = (INDEX_MCP, MIDDLE_MCP, RING_MCP, PINKY_MCP)
PALM_POINTS = (WRIST, INDEX_MCP, MIDDLE_MCP, RING_MCP, PINKY_MCP)

# --- MediaPipe pose landmark indices (33-point model) ---
POSE_LEFT_SHOULDER = 11
POSE_RIGHT_SHOULDER = 12


@dataclass
class Hand:
    """One detected hand: 21 landmarks in pixel coords (x, y) plus relative depth (z)."""

    handedness: str            # "Left" or "Right", from the image's perspective
    points: np.ndarray         # shape (21, 3): x_px, y_px, z (MediaPipe relative depth)

    @property
    def wrist(self) -> np.ndarray:
        return self.points[WRIST, :2]

    @property
    def center(self) -> np.ndarray:
        """Palm-center proxy: mean of wrist + finger MCPs (x, y in px). Stable pivot for motion."""
        return self.points[list(PALM_POINTS), :2].mean(axis=0)

    def to_dict(self) -> dict:
        return {"handedness": self.handedness, "points": self.points.tolist()}

    @classmethod
    def from_dict(cls, d: dict) -> "Hand":
        return cls(handedness=d["handedness"], points=np.asarray(d["points"], dtype=float))


@dataclass
class Frame:
    """One timestamped capture: any detected hands + shoulder points for scale normalization."""

    t: float                                          # timestamp, seconds (monotonic)
    width: int                                        # source image width (px)
    height: int                                       # source image height (px)
    hands: list[Hand] = field(default_factory=list)   # 0..2 hands, in detection order
    left_shoulder: Optional[np.ndarray] = None        # (x, y) px
    right_shoulder: Optional[np.ndarray] = None       # (x, y) px

    @property
    def shoulder_width(self) -> Optional[float]:
        """Scale reference in px. None if pose/shoulders weren't detected."""
        if self.left_shoulder is None or self.right_shoulder is None:
            return None
        w = float(np.linalg.norm(self.left_shoulder - self.right_shoulder))
        return w if w > 1e-6 else None

    def hand(self, handedness: str) -> Optional[Hand]:
        """First hand matching 'Left'/'Right', or None."""
        for h in self.hands:
            if h.handedness == handedness:
                return h
        return None

    @property
    def has_both_hands(self) -> bool:
        return len(self.hands) >= 2

    @property
    def is_complete(self) -> bool:
        """Enough data to evaluate a two-handed sign: both hands + a shoulder scale."""
        return self.has_both_hands and self.shoulder_width is not None

    def to_dict(self) -> dict:
        return {
            "t": self.t,
            "width": self.width,
            "height": self.height,
            "hands": [h.to_dict() for h in self.hands],
            "left_shoulder": None if self.left_shoulder is None else self.left_shoulder.tolist(),
            "right_shoulder": None if self.right_shoulder is None else self.right_shoulder.tolist(),
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Frame":
        return cls(
            t=d["t"],
            width=d["width"],
            height=d["height"],
            hands=[Hand.from_dict(h) for h in d["hands"]],
            left_shoulder=None if d.get("left_shoulder") is None else np.asarray(d["left_shoulder"], float),
            right_shoulder=None if d.get("right_shoulder") is None else np.asarray(d["right_shoulder"], float),
        )


def normalized_distance(p1, p2, shoulder_width: float) -> float:
    """Euclidean distance between two pixel points, as a ratio of shoulder width."""
    return float(np.linalg.norm(np.asarray(p1, float) - np.asarray(p2, float)) / shoulder_width)


class RollingBuffer:
    """A fixed time-window (seconds) of Frames. The basis for every movement check.

    Old frames are evicted as new ones arrive, so the buffer always holds roughly the last
    `window_seconds` of motion — never a single frame. Movement detectors read start/end and
    iterate the window; they never look at just the current frame.
    """

    def __init__(self, window_seconds: float = 2.0):
        self.window_seconds = window_seconds
        self._frames: deque[Frame] = deque()

    def add(self, frame: Frame) -> None:
        self._frames.append(frame)
        cutoff = frame.t - self.window_seconds
        while self._frames and self._frames[0].t < cutoff:
            self._frames.popleft()

    def clear(self) -> None:
        self._frames.clear()

    def __len__(self) -> int:
        return len(self._frames)

    def __iter__(self) -> Iterator[Frame]:
        return iter(self._frames)

    @property
    def frames(self) -> list[Frame]:
        return list(self._frames)

    @property
    def start(self) -> Optional[Frame]:
        return self._frames[0] if self._frames else None

    @property
    def end(self) -> Optional[Frame]:
        return self._frames[-1] if self._frames else None

    @property
    def duration(self) -> float:
        """Seconds spanned by the buffer (0 until at least two frames are present)."""
        return (self._frames[-1].t - self._frames[0].t) if len(self._frames) >= 2 else 0.0


class HandStabilizer:
    """Bridge brief hand-detection dropouts by carrying a recently-seen hand forward.

    MediaPipe intermittently loses a hand for a frame or two — closed fists especially, and worse
    in poor light. For LIVE PLAY, hold the last-seen hand for up to `hold_seconds` so a momentary
    miss doesn't reset location/handshape scoring (the flicker users feel as "it keeps dropping").

    Do NOT use this when recording fixtures — training data must stay raw/honest. The carried
    hand keeps its last position, which is exactly right for the held-still non-dominant fist and
    harmless for the brief gaps it bridges.
    """

    def __init__(self, hold_seconds: float = 0.3):
        self.hold_seconds = hold_seconds
        self._last: dict[str, tuple[float, Hand]] = {}

    def reset(self) -> None:
        self._last.clear()

    def stabilize(self, frame: Frame) -> Frame:
        """Update memory with the frame's real hands, then re-add any recently-seen missing ones."""
        present = set()
        for h in frame.hands:
            self._last[h.handedness] = (frame.t, h)
            present.add(h.handedness)
        for handedness, (t_seen, hand) in self._last.items():
            if handedness not in present and (frame.t - t_seen) <= self.hold_seconds:
                frame.hands.append(hand)
        return frame

File: web/src/engine/landmarks.ts
import { norm2d, mean as arrMean } from './math-utils';

export const WRIST = 0;
export const THUMB_TIP = 4;
export const INDEX_MCP = 5;
export const INDEX_TIP = 8;
export const MIDDLE_MCP = 9;
export const MIDDLE_TIP = 12;
export const RING_MCP = 13;
export const RING_TIP = 16;
export const PINKY_MCP = 17;
export const PINKY_TIP = 20;
export const FINGERTIPS = [THUMB_TIP, INDEX_TIP, MIDDLE_TIP, RING_TIP, PINKY_TIP] as const;
export const MCPS = [INDEX_MCP, MIDDLE_MCP, RING_MCP, PINKY_MCP] as const;
export const PALM_POINTS = [WRIST, INDEX_MCP, MIDDLE_MCP, RING_MCP, PINKY_MCP] as const;

export const POSE_NOSE = 0;
export const POSE_MOUTH_LEFT = 9;
export const POSE_MOUTH_RIGHT = 10;
export const POSE_LEFT_SHOULDER = 11;
export const POSE_RIGHT_SHOULDER = 12;

export interface Hand {
  handedness: string;
  points: number[][]; // (21, 3)
}

export function handWrist(h: Hand): number[] {
  return h.points[WRIST].slice(0, 2);
}

export function handCenter(h: Hand): number[] {
  const xs: number[] = [];
  const ys: number[] = [];
  for (const idx of PALM_POINTS) {
    xs.push(h.points[idx][0]);
    ys.push(h.points[idx][1]);
  }
  return [arrMean(xs), arrMean(ys)];
}

export interface Frame {
  t: number;
  width: number;
  height: number;
  hands: Hand[];
  leftShoulder: number[] | null;
  rightShoulder: number[] | null;
  mouth: number[] | null;
}

export function frameShoulderWidth(f: Frame): number | null {
  if (!f.leftShoulder || !f.rightShoulder) return null;
  const w = norm2d(f.leftShoulder, f.rightShoulder);
  return w > 1e-6 ? w : null;
}

export function frameHand(f: Frame, handedness: string): Hand | null {
  for (const h of f.hands) {
    if (h.handedness === handedness) return h;
  }
  return null;
}

export function frameHasBothHands(f: Frame): boolean {
  return f.hands.length >= 2;
}

export function frameIsComplete(f: Frame): boolean {
  return frameHasBothHands(f) && frameShoulderWidth(f) !== null;
}

export function normalizedDistance(
  p1: number[],
  p2: number[],
  shoulderWidth: number
): number {
  return norm2d(p1, p2) / shoulderWidth;
}

export function frameFromDict(d: {
  t: number;
  width: number;
  height: number;
  hands: { handedness: string; points: number[][] }[];
  left_shoulder?: number[] | null;
  right_shoulder?: number[] | null;
  mouth?: number[] | null;
}): Frame {
  return {
    t: d.t,
    width: d.width,
    height: d.height,
    hands: d.hands.map((h) => ({ handedness: h.handedness, points: h.points })),
    leftShoulder: d.left_shoulder ?? null,
    rightShoulder: d.right_shoulder ?? null,
    mouth: d.mouth ?? null,
  };
}

export class RollingBuffer {
  windowSeconds: number;
  private _frames: Frame[] = [];

  constructor(windowSeconds = 2.0) {
    this.windowSeconds = windowSeconds;
  }

  add(frame: Frame): void {
    this._frames.push(frame);
    const cutoff = frame.t - this.windowSeconds;
    while (this._frames.length > 0 && this._frames[0].t < cutoff) {
      this._frames.shift();
    }
  }

  clear(): void {
    this._frames = [];
  }

  get length(): number {
    return this._frames.length;
  }

  get frames(): Frame[] {
    return [...this._frames];
  }

  [Symbol.iterator](): Iterator<Frame> {
    let i = 0;
    const frames = this._frames;
    return {
      next() {
        if (i < frames.length) return { value: frames[i++], done: false };
        return { value: undefined as never, done: true };
      },
    };
  }

  get start(): Frame | null {
    return this._frames.length > 0 ? this._frames[0] : null;
  }

  get end(): Frame | null {
    return this._frames.length > 0 ? this._frames[this._frames.length - 1] : null;
  }

  get duration(): number {
    if (this._frames.length < 2) return 0;
    return this._frames[this._frames.length - 1].t - this._frames[0].t;
  }
}

export class HandStabilizer {
  holdSeconds: number;
  private _last: Map<string, { t: number; hand: Hand }> = new Map();

  constructor(holdSeconds = 0.3) {
    this.holdSeconds = holdSeconds;
  }

  reset(): void {
    this._last.clear();
  }

  stabilize(frame: Frame): Frame {
    const present = new Set<string>();
    for (const h of frame.hands) {
      this._last.set(h.handedness, { t: frame.t, hand: h });
      present.add(h.handedness);
    }
    for (const [handedness, { t, hand }] of this._last) {
      if (!present.has(handedness) && frame.t - t <= this.holdSeconds) {
        frame.hands.push(hand);
      }
    }
    return frame;
  }
}
